Implicit modifier crashes the typer

The following snippet fails to typecheck, but also crashes the typer:

```
object Foo {
        def foo(): Int = {
                val f: implicit Int => Int = (implicit x: Int) => 2 * x
                f(2)
        }

}
```
Results:

```
Building tests...done
-- [E039] Syntax Error: ../test/kk.scala:6:40 ----------------------------------
6 |		val f: implicit Int => Int = (implicit x: Int) => 2 * x
  |		                                        ^
  |		                                        '=>' expected, but ':' found

exception occurred while typechecking ../test/kk.scala

exception occurred while compiling ../test/kk.scala
Exception in thread "main" java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedTypeTree$1.apply(Typer.scala:1010)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedTypeTree$1.apply(Typer.scala:995)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedTypeTree(Typer.scala:994)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1546)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedType(Typer.scala:1643)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:1176)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:1173)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:1173)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1511)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedAppliedTypeTree$1.dotty$tools$dotc$typer$Typer$$anonfun$$typedArg$1(Typer.scala:1078)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedAppliedTypeTree$1$$anonfun$21.apply(Typer.scala:1080)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedAppliedTypeTree$1$$anonfun$21.apply(Typer.scala:1080)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.zipWithConserve$extension(Decorators.scala:108)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedAppliedTypeTree$1.apply(Typer.scala:1080)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedAppliedTypeTree$1.apply(Typer.scala:1051)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedAppliedTypeTree(Typer.scala:1051)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1551)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedFunction$1.apply(Typer.scala:675)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedFunction$1.apply(Typer.scala:671)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedFunction(Typer.scala:671)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1535)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Namer$$anonfun$typedAheadType$1.apply(Namer.scala:931)
	at dotty.tools.dotc.typer.Namer$$anonfun$typedAheadType$1.apply(Namer.scala:931)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:924)
	at dotty.tools.dotc.typer.Namer.typedAheadType(Namer.scala:931)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1096)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:729)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:777)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:755)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:219)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:197)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:336)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:1479)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1504)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Namer$$anonfun$typedAheadExpr$1.apply(Namer.scala:934)
	at dotty.tools.dotc.typer.Namer$$anonfun$typedAheadExpr$1.apply(Namer.scala:934)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:924)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:934)
	at dotty.tools.dotc.typer.Namer$$anonfun$completeParams$1.apply(Namer.scala:946)
	at dotty.tools.dotc.typer.Namer$$anonfun$completeParams$1.apply(Namer.scala:946)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at dotty.tools.dotc.typer.Namer.completeParams(Namer.scala:946)
	at dotty.tools.dotc.typer.Namer$$anonfun$defDefSig$3.apply(Namer.scala:1131)
	at dotty.tools.dotc.typer.Namer$$anonfun$defDefSig$3.apply(Namer.scala:1131)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1131)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:733)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:777)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:755)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:219)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:197)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:336)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:1479)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1504)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1609)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:598)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedBlock$1.apply(Typer.scala:601)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedBlock$1.apply(Typer.scala:600)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:600)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1533)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedFunction$1.apply(Typer.scala:789)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedFunction$1.apply(Typer.scala:671)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedFunction(Typer.scala:671)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1535)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:1237)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:1219)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:1219)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1514)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1609)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:598)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedBlock$1.apply(Typer.scala:601)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedBlock$1.apply(Typer.scala:600)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:600)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1533)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedFunction$1.apply(Typer.scala:789)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedFunction$1.apply(Typer.scala:671)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedFunction(Typer.scala:671)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1535)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.makeImplicitFunction(Typer.scala:1582)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1569)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:1179)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:1173)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:1173)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1511)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1609)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:598)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedBlock$1.apply(Typer.scala:601)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedBlock$1.apply(Typer.scala:600)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:600)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1533)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:1237)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:1219)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:1219)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1514)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1609)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1319)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1267)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1267)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1517)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1571)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1609)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:1432)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:1419)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1419)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1556)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1572)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1587)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1585)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:140)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:57)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1585)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1641)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$typeCheck$1.apply$mcV$sp(FrontEnd.scala:64)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:32)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:62)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:93)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:93)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:93)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:82)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:79)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:79)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:67)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:67)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:67)
	at dotty.tools.dotc.Run.compileSources(Run.scala:64)
	at dotty.tools.dotc.Run.compile(Run.scala:48)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at dotty.tools.dotc.Driver.process(Driver.scala:93)
	at dotty.tools.dotc.Driver.process(Driver.scala:105)
	at dotty.tools.dotc.Driver.main(Driver.scala:132)
	at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -380,7 +380,7 @@ object desugar {
       def enumTagMeths = if (isEnumCase) enumTagMeth(CaseKind.Class)._1 :: Nil else Nil
       def copyMeths = {
         def isRepeated(tree: Tree): Boolean = tree match {
-          case PostfixOp(_, Ident(nme.raw.STAR)) => true
+          case PostfixOp(_, Ident(tpnme.raw.STAR)) => true
           case ByNameTypeTree(tree1) => isRepeated(tree1)
           case _ => false
         }
@@ -1051,7 +1051,7 @@ object desugar {
         else // l.op(r), or val x = r; l.op(x), plus handle named args specially
           makeBinop(l, op, r)
       case PostfixOp(t, op) =>
-        if ((ctx.mode is Mode.Type) && !op.isBackquoted && op.name == nme.raw.STAR) {
+        if ((ctx.mode is Mode.Type) && !op.isBackquoted && op.name == tpnme.raw.STAR) {
           val seqType = if (ctx.compilationUnit.isJava) defn.ArrayType else defn.SeqType
           Annotated(
             AppliedTypeTree(ref(seqType), t),

--- a/compiler/src/dotty/tools/dotc/ast/untpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/untpd.scala
@@ -367,7 +367,7 @@ object untpd extends Trees.Instance[Untyped] with UntypedTreeInfo {
    *  parameter, the reference will be a repeated argument.
    */
   def refOfDef(tree: MemberDef)(implicit ctx: Context) = tree match {
-    case ValDef(_, PostfixOp(_, Ident(nme.raw.STAR)), _) => repeated(Ident(tree.name))
+    case ValDef(_, PostfixOp(_, Ident(tpnme.raw.STAR)), _) => repeated(Ident(tree.name))
     case _ => Ident(tree.name)
   }
 

--- a/compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala
@@ -447,7 +447,7 @@ object JavaParsers {
       if (in.token == DOTDOTDOT) {
         in.nextToken()
         t = atPos(t.pos.start) {
-          PostfixOp(t, Ident(nme.raw.STAR))
+          PostfixOp(t, Ident(tpnme.raw.STAR))
         }
       }
       atPos(start, in.offset) {

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -916,7 +916,7 @@ object Parsers {
       val t = toplevelTyp()
       if (isIdent(nme.raw.STAR)) {
         in.nextToken()
-        atPos(startOffset(t)) { PostfixOp(t, Ident(nme.raw.STAR)) }
+        atPos(startOffset(t)) { PostfixOp(t, Ident(tpnme.raw.STAR)) }
       } else t
     }
 
@@ -1322,7 +1322,7 @@ object Parsers {
      *                    |  `(' [ExprsInParens `,'] PostfixExpr `:' `_' `*' ')'
      *
      *  Special treatment for arguments of primary class constructor
-     *  annotations. All empty argument lists `(` `)` following the first 
+     *  annotations. All empty argument lists `(` `)` following the first
      *  get represented as `List(ParamNotArg)` instead of `Nil`, indicating that
      *  the token sequence should be interpreted as an empty parameter clause
      *  instead. `ParamNotArg` can also be produced when parsing the first
@@ -2047,7 +2047,7 @@ object Parsers {
 
     private def checkVarArgsRules(vparamss: List[List[untpd.ValDef]]): List[untpd.ValDef] = {
       def isVarArgs(tpt: Trees.Tree[Untyped]): Boolean = tpt match {
-        case PostfixOp(_, op) if op.name == nme.raw.STAR => true
+        case PostfixOp(_, op) if op.name == tpnme.raw.STAR => true
         case _ => false
       }
 

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -1056,8 +1056,8 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
             throw ex
         }
       case _ =>
-        assert(isFullyDefined(pt, ForceDegree.none))
-        tree.withType(pt)
+        tree.withType(
+          if (isFullyDefined(pt, ForceDegree.none)) pt else UnspecifiedErrorType)
     }
   }
 
@@ -1124,7 +1124,8 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
                 tparam.ensureCompleted()
               case _ =>
             }
-          typed(desugaredArg, argPt)
+          if (desugaredArg.isType) typed(desugaredArg, argPt)
+          else desugaredArg.withType(UnspecifiedErrorType)
         }
         args.zipWithConserve(tparams)(typedArg(_, _)).asInstanceOf[List[Tree]]
       }

--- a/tests/neg/i2514.scala
+++ b/tests/neg/i2514.scala
@@ -1,0 +1,10 @@
+object Foo {
+  def foo(): Int = {
+    val f: implicit Int => Int = (implicit x: Int) => 2 * x // error // error
+    f(2)
+  }
+
+  val f = (implicit x: Int) => x // error // error
+
+  ((implicit x: Int) => x): (implicit Int => Int) // error // error // error
+}

--- a/tests/pos/Result.scala
+++ b/tests/pos/Result.scala
@@ -1,0 +1,17 @@
+import scala.util.control.NonFatal
+object p {
+
+  enum Result[+T, +E] {
+    case OK [T](x: T) extends Result[T, Nothing]
+    case Err[E](e: E) extends Result[Nothing, E]
+  }
+
+  type Try[T] = Result[T, Throwable]
+  object Try {
+    def apply[T](x: => T): Try[T] =
+      try Result.OK(x)
+      catch {
+        case NonFatal(ex) => Result.Err(ex)
+      }
+  }
+}

